### PR TITLE
MAINT: Move `np.dtype.name.__get__` to python

### PR DESCRIPTION
**Pull request by eric-wieser**
This shares a lot of code with part of `dtype.__repr__`, and it's helpful to get it all in one place.

This fixes possible but unlikely segfaults if `PyUString_ConcatAndDel` fails.

Follows up on some remarks left in #10602, hopefully makes it easier to solve problems in #10151.

### CODE REVIEW

File: numpy/core/src/multiarray/descriptor.c [hunk 1]
```diff
@@ -1859,72 +1859,17 @@ arraydescr_protocol_typestr_get(PyArray_Descr *self)
...
-    int prefix_len;
-    int suffix_len;
-
-    if (PyTypeNum_ISUSERDEF(self->type_num)) {
-        s = strrchr(typeobj->tp_name, '.');
-        if (s == NULL) {
-            res = PyUString_FromString(typeobj->tp_name);
-        }
-        else {
-            res = PyUString_FromStringAndSize(s + 1, strlen(s) - 1);
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/descriptor.c]: This entire block is just `typeobj.__name__`

File: numpy/core/_dtype.py [hunk 2]
```diff
@@ -292,3 +298,26 @@ def _struct_repr(dtype):
...
+    # Builtin classes are documented as returning a "bit name"
+    name = dtype.type.__name__
+
+    # handle bool_, str_, etc
+    if name[-1] == '_':
+        name = name[:-1]
+
+    # append bit counts to str, unicode, and void
+    if np.issubdtype(dtype, np.flexible) and not _isunsized(dtype):
+        name += "{}".format(dtype.itemsize * 8)
```
Comment (eric-wieser) [inline numpy/core/_dtype.py]: @charris: As predicted, the assumption that a byte is 8 bits is pretty ingrained into our dtype string handling

File: numpy/core/src/multiarray/descriptor.c [hunk 3]
```diff
@@ -1859,72 +1859,17 @@ arraydescr_protocol_typestr_get(PyArray_Descr *self)
...
 
 static PyObject *
-arraydescr_typename_get(PyArray_Descr *self)
+arraydescr_name_get(PyArray_Descr *self)
 {
-    static const char np_prefix[] = "numpy.";
-    const int np_prefix_len = sizeof(np_prefix) - 1;
-    PyTypeObject *typeobj = self->typeobj;
+    /* let python handle this */
+    PyObject *_numpy_dtype;
```
Comment (charris) [inline numpy/core/src/multiarray/descriptor.c]: Shouldn't this be `static PyObject *_numpy_dtype = NULL;`  It needs to stay around and needs to be initialized.

Comment (eric-wieser) [reply numpy/core/src/multiarray/descriptor.c]: I don't think it does need to stay around, and it's initialized at its first use below.

This matches `dtype.__repr__`, in that both do not bother to cache the module or the function

**Pull request merged**